minifi/minifi-bootstrap/src/main/java/org/apache/nifi/minifi/bootstrap/service/MiNiFiConfigurationChangeListener.java

```diff
@@ -0,0 +1,208 @@
+            }
+
+            File swapConfigFile = bootstrapFileProvider.getSwapFile();
+            logger.info("Persisting old configuration to {}", swapConfigFile.getAbsolutePath());
+
+            try (FileInputStream configFileInputStream = new FileInputStream(configFile)) {
+                Files.copy(configFileInputStream, swapConfigFile.toPath(), REPLACE_EXISTING);
+            }
+
+            persistBackNonFlowSectionsFromOriginalSchema(bufferedConfigOs.toByteArray(), bootstrapProperties, configFile);
+
+            // Create an input stream to feed to the config transformer
+            try (FileInputStream newConfigIs = new FileInputStream(configFile)) {
+
+                try {
+                    String confDir = bootstrapProperties.getProperty(CONF_DIR_KEY);
+
+                    try {
+                        logger.info("Performing transformation for input and saving outputs to {}", confDir);
+                        ByteBuffer tempConfigFile = generateConfigFiles(newConfigIs, confDir, bootstrapFileProvider.getBootstrapProperties());
+                        runner.getConfigFileReference().set(tempConfigFile.asReadOnlyBuffer());
+
+                        try {
+                            logger.info("Reloading instance with new configuration");
+                            restartInstance();
+                        } catch (Exception e) {
+                            logger.debug("Transformation of new config file failed after transformation into Flow.xml and nifi.properties, reverting.");
+                            try (FileInputStream swapConfigFileStream = new FileInputStream(swapConfigFile)) {
+                                ByteBuffer resetConfigFile = generateConfigFiles(swapConfigFileStream, confDir, bootstrapFileProvider.getBootstrapProperties());
+                                runner.getConfigFileReference().set(resetConfigFile.asReadOnlyBuffer());
+                            }
+                            throw e;
+                        }
+                    } catch (Exception e) {
+                        logger.debug("Transformation of new config file failed after replacing original with the swap file, reverting.");
+                        try (FileInputStream swapConfigFileStream = new FileInputStream(swapConfigFile)) {
+                            Files.copy(swapConfigFileStream, configFile.toPath(), REPLACE_EXISTING);
+                        }
+                        throw e;
+                    }
+                } catch (Exception e) {
+                    logger.debug("Transformation of new config file failed after swap file was created, deleting it.");
+                    if (!swapConfigFile.delete()) {
+                        logger.warn("The swap file failed to delete after a failed handling of a change. It should be cleaned up manually.");
+                    }
+                    throw e;
+                }
+            }
```

Comment:
This set of try-catch blocks is nested deeply, it would be helpful to refactor to separate methods.

Reply:
Restructured and moved some to private functions. It could be refactored even further but didn't wanted to "overdo" it as there is upcoming work where we would pass the flow content to flowservice directly which will render most of the code here obsolete (and all of this will be removed).